File: sleepy-test/python-unittest/AsyncRequestQueueTests_AsyncRequestQueueWithBlockBehavior.py
import unittest
import threading
import time

class TestAsyncRequestQueue(unittest.TestCase):
    def test_async_request_queue_with_block_behavior(self):
        queue = AsyncRequestQueue(10, AsyncTargetWrapperOverflowAction.BLOCK)
        producer_finished = threading.Event()
        pushing_event = [0]

        def producer_thread():
            for i in range(1000):
                log_event = LogEventInfo.create_null_event().with_continuation(lambda ex: None)
                log_event.log_event.message = f"msg{i}"
                pushing_event[0] = i
                queue.enqueue(log_event)
            producer_finished.set()

        threading.Thread(target=producer_thread).start()

        total = 0
        while total < 500:
            left = 500 - total
            log_event_infos = queue.dequeue_batch(left)
            got = len(log_event_infos)
            self.assertTrue(got <= queue.request_limit)
            total += got

        time.sleep(0.5)
        self.assertEqual(510, pushing_event[0])

        queue.dequeue_batch(1)
        total += 1
        time.sleep(0.5)
        self.assertEqual(511, pushing_event[0])

        while total < 1000:
            left = 1000 - total
            log_event_infos = queue.dequeue_batch(left)
            got = len(log_event_infos)
            self.assertTrue(got <= queue.request_limit)
            total += got

        producer_finished.wait()

File: sleepy-test/CSharpXUnit/CSharpXUnit.Tests/Tests/AsyncRequestQueueTests_AsyncRequestQueueWithBlockBehavior.cs
using Xunit;

namespace nlog
{
    public class AsyncRequestQueueTests
    {
        [Fact]
        public void AsyncRequestQueueWithBlockBehavior()
        {
            {
                var queue = new AsyncRequestQueue(10, AsyncTargetWrapperOverflowAction.Block);
                ManualResetEvent producerFinished = new ManualResetEvent(false);
                int pushingEvent = 0;
                ThreadPool.QueueUserWorkItem(s =>
                {
                    // producer thread
                    for (int i = 0; i < 1000; ++i)
                    {
                        AsyncLogEventInfo logEvent = LogEventInfo.CreateNullEvent().WithContinuation(ex =>
                        {
                        });
                        logEvent.LogEvent.Message = "msg" + i;
                        // Console.WriteLine("Pushing event {0}", i);
                        pushingEvent = i;
                        queue.Enqueue(logEvent);
                    }
            
                    producerFinished.Set();
                });
                // consumer thread
                AsyncLogEventInfo[] logEventInfos;
                int total = 0;
                while (total < 500)
                {
                    int left = 500 - total;
                    logEventInfos = queue.DequeueBatch(left);
                    int got = logEventInfos.Length;
                    Assert.True(got <= queue.RequestLimit);
                    total += got;
                }
            
                Thread.Sleep(500);
                // producer is blocked on trying to push event #510
                Assert.Equal(510, pushingEvent);
                queue.DequeueBatch(1);
                total++;
                Thread.Sleep(500);
                // producer is now blocked on trying to push event #511
                Assert.Equal(511, pushingEvent);
                while (total < 1000)
                {
                    int left = 1000 - total;
                    logEventInfos = queue.DequeueBatch(left);
                    int got = logEventInfos.Length;
                    Assert.True(got <= queue.RequestLimit);
                    total += got;
                }
            
                // producer should now finish
                producerFinished.WaitOne();
            }
        }
    }
}
